EIP 3529: alternative refund reduction (#3529)
ethereum/EIPs: EIPS/eip-3529.md

---
eip: 3529
title: Reduction in refunds
author: Vitalik Buterin (@vbuterin), Martin Swende (@holiman)
discussions-to: https://ethereum-magicians.org/t/eip-3529-reduction-in-refunds-alternative-to-eip-3298-and-3403-that-better-preserves-existing-clearing-incentives/6097
status: Draft
type: Standards Track
category: Core
created: 2021-04-22
---

## Simple Summary

> Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE to a lower level where the refunds are still substantial, but they are no longer high enough for current "exploits" of the refund mechanism to be viable.

Reviewer: Recommend shortening this up.  Think email subject line or forum thread title or GitHub PR title.
```suggestion
Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE.
```
The latter part of this sentence should be left for the motivation section.

* Refunds give rise to GasToken. GasToken has benefits in moving gas space from low-fee periods to high-fee periods, but it also has downsides to the network, particularly in exacerbating state size (as state slots are effectively used as a "battery" to save up gas) and inefficiently clogging blockchain gas usage
* Refunds increase block size variance. The theoretical maximum amount of actual gas consumed in a block is nearly twice the on-paper gas limit (as refunds add gas space for subsequent transactions in a block, though refunds are capped at 50% of a transaction's gas used). This is not fatal, but is still undesirable, especially given that refunds can be used to maintain 2x usage spikes for far longer than EIP 1559 can.

## Specification

### Parameters

| Constant | Value |
| - | - |
| `FORK_BLOCK` | TBD |
| `NEW_MAX_REFUND_QUOTIENT` | 5 |

For blocks where `block.number >= FORK_BLOCK`, the following changes apply.

1. Remove the `SELFDESTRUCT` refund.

> 2. Replace `SSTORE_CLEARS_SCHEDULE` with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))

Author: ```suggestion
2. Replace `SSTORE_CLEARS_SCHEDULE` (as defined in [EIP-2200](./eip-2929.md)) with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))
```

> 2. Replace `SSTORE_CLEARS_SCHEDULE` (as defined in [EIP-2200](./eip-2929.md)) with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))

Author: ```suggestion
2. Replace `SSTORE_CLEARS_SCHEDULE` (as defined in [EIP-2200](./eip-2200.md)) with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of [EIP-2929](./eip-2929.md) + [EIP-2930](./eip-2930.md))
```

---
eip: 3529
title: Reduction in refunds
author: Vitalik Buterin (@vbuterin), Martin Swende (@holiman)

> discussions-to: https://ethereum-magicians.org/t/eip-3298-removal-of-refunds/5430

Author: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-3529-reduction-in-refunds-alternative-to-eip-3298-and-3403-that-better-preserves-existing-clearing-incentives/6097
```

status: Draft
type: Standards Track
category: Core
created: 2021-04-22
---

## Simple Summary

> Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE.

Author: ```suggestion
Remove gas refunds for SELFDESTRUCT, and reduce gas refunds for SSTORE to a lower level where the refunds are still substantial, but they are no longer high enough for current "exploits" of the refund mechanism to be viable.
```

* Refunds increase block size variance. The theoretical maximum amount of actual gas consumed in a block is nearly twice the on-paper gas limit (as refunds add gas space for subsequent transactions in a block, though refunds are capped at 50% of a transaction's gas used). This is not fatal, but is still undesirable, especially given that refunds can be used to maintain 2x usage spikes for far longer than EIP 1559 can.

## Specification

### Parameters

| Constant | Value |
| - | - |
| `FORK_BLOCK` | TBD |
| `NEW_MAX_REFUND_QUOTIENT` | 5 |

For blocks where `block.number >= FORK_BLOCK`, the following changes apply.

1. Remove the `SELFDESTRUCT` refund.
2. Replace `SSTORE_CLEARS_SCHEDULE` with `SSTORE_RESET_GAS + ACCESS_LIST_STORAGE_KEY_COST` (4,800 gas as of EIP 2929 + 2930)
3. Reduce the max gas refunded after a transaction to `gas_used // NEW_MAX_REFUND_QUOTIENT`

## Rationale

> In [EIP 2200](https://eips.ethereum.org/EIPS/eip-2200#specification), three cases for refunds were introduced:

Reviewer: ```suggestion
In [EIP-2200](./eip-2200.md#specification), three cases for refunds were introduced:
```

Gas tokens will become valueless. DeFi arbitrage bots, which today frequently use either established gas token schemes or a custom alternative to reduce on-chain costs, would benefit from rewriting their code to remove calls to these no-longer-functional gas storage mechanisms.

However, fully preserving refunds in the `new = original = 0 != current` case, and keeping _some_ refund in the other `nonzero -> zero` cases, ensures that a few key use cases that receive (and deserve) favorable gas cost treatment continue to do so. For example, `zero -> nonzero -> zero` storage set patterns continue to cost only ~100 gas. Two important examples of such patterns include:

* Anti-reentrancy locks (typically flipped from 0 to 1 right before a child call begins, and then flipped back to 0 when the child call ends) 
* ERC20 approve-and-send (the "approved value" goes from zero to nonzero when the token transfer is approved, and then back to zero when the token transfer processes)

### Effect on storage clearing incentives

> A criticism of earlier refund removal EIPs ([EIP 3298](https://eips.ethereum.org/EIPS/eip-3298) and [EIP 3403](https://eips.ethereum.org/EIPS/eip-3403)) is that these EIPs fully remove the incentive to set a value to zero, encouraging users to not fully clear a storage slot if they expect even the smallest probability that they will want to use that storage slot again.

Reviewer: ```suggestion
A criticism of earlier refund removal EIPs ([EIP-3298](./eip-3298.md) and [EIP-3403](./eip-3403.md)) is that these EIPs fully remove the incentive to set a value to zero, encouraging users to not fully clear a storage slot if they expect even the smallest probability that they will want to use that storage slot again.
```

Reviewer: EIP-3298 and 3403 are unlikely to be finalized before this one, which means referencing them will block this EIP from moving to final.  Recommend removing references to them and instead just vaguely discuss alternative proposals directly here like "Another option would be to do X, but that would have problem Y."

For example, if you have 1 unit of an ERC20 token and you are giving away or selling your entire balance, you could instead only give away 0.999999 units and leave the remainder behind. If you ever decide to re-acquire more of that token with the same account in the future, you would only have to pay 5000 gas (2100 for the read + 2900 for nonzero-to-nonzero set) for the SSTORE instead of 22100 (20000 for the zero-to-nonzero set). Today, this is counterbalanced by the 15000 refund for clearing, so you only have an incentive to do this if you are more than `15000 / 17100 = 87.7%` sure that you will use the slot again; with EIP 3298 or EIP 3403 the counterbalancing incentive would not exist, so setting to nonzero is better if your chance of using the slot again is _any_ value greater than 0%.

> If this EIP were to be implemented, a refund of 4800 gas would remain, so there would only be an incentive to keep a storage slot nonzero if you expect a probability of more than `4800 / 17100 = 28.1%` that you will use that slot again. This is not perfect, but it is likely higher than the average person's expectations of later re-acquiring a token with the same address if they clear their entire balance of it.

Reviewer: EIPs should be written for a target audience of future readers after it has been decided that it is desirable.
```suggestion
A refund of 4800 gas remains, so there is only be an incentive to keep a storage slot nonzero if you expect a probability of more than `4800 / 17100 = 28.1%` that you will use that slot again. This is not perfect, but it is likely higher than the average person's expectations of later re-acquiring a token with the same address if they clear their entire balance of it.
```

| -- | -- | -- | -- | -- | -- | -- | -- | 
| `0x60006000556000600055` | 212 | 0| 0 | 0 |  0 |  |  212 |
| `0x60006000556001600055` | 20112 | 0| 0 | 0 |  1 |  |  20112 |
| `0x60016000556000600055` | 20112 | 19900| 0 | 1 |  0 |  |  212 |
| `0x60016000556002600055` | 20112 | 0| 0 | 1 |  2 |  |  20112 |
| `0x60016000556001600055` | 20112 | 0| 0 | 1 |  1 |  |  20112 |
| `0x60006000556000600055` | 3012 | 4800| 1 | 0 |  0 |  |  -1788 |
| `0x60006000556001600055` | 3012 | 2800| 1 | 0 |  1 |  |  212 |
| `0x60006000556002600055` | 3012 | 0| 1 | 0 |  2 |  |  3012 |
| `0x60026000556000600055` | 3012 | 4800| 1 | 2 |  0 |  |  -1788 |
| `0x60026000556003600055` | 3012 | 0| 1 | 2 |  3 |  |  3012 |
| `0x60026000556001600055` | 3012 | 2800| 1 | 2 |  1 |  |  212 |
| `0x60026000556002600055` | 3012 | 0| 1 | 2 |  2 |  |  3012 |
| `0x60016000556000600055` | 3012 | 4800| 1 | 1 |  0 |  |  -1788 |
| `0x60016000556002600055` | 3012 | 0| 1 | 1 |  2 |  |  3012 |
| `0x60016000556001600055` | 212 | 0| 1 | 1 |  1 |  |  212 |
| `0x600160005560006000556001600055` | 40118 | 19900| 0 | 1 |  0 |  1 |  20218 |
| `0x600060005560016000556000600055` | 5918 | 7600| 1 | 0 |  1 |  0 |  -1682 |

> ## The mutex usecase
> 
> There are two typical ways to implement mutexes: '0-1-0' and '1-2-1. Let's see how they differ
> 
> - '0-1-0':
>   - Istanbul: 1612
>   - Berlin: 212
>   - EIP-3403: 1112
>   - This EIP: 212
> - '1-2-1':
>   - Istanbul: 1612
>   - Berlin: 212
>   - EIP-3403: 3012
>   - This EIP: 212
> 

Reviewer: This isn't a valid EIP section.  Consider converting this into a test case, or putting it as a subheading in the Rationale section.
```suggestion
```
